src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -767,28 +843,54 @@ private async Task<Dictionary<string, ConfigurationSetting>> LoadSelectedKeyValu
                     LabelFilter = LabelFilter.Null
                 };
 
+                List<MatchConditions> matchConditions = new List<MatchConditions>();
+
                 await CallWithRequestTracing(async () =>
                 {
-                    await foreach (ConfigurationSetting setting in client.GetConfigurationSettingsAsync(selector, cancellationToken).ConfigureAwait(false))
+                    AsyncPageable<ConfigurationSetting> pageableSettings = client.GetConfigurationSettingsAsync(selector, cancellationToken);
+
+                    await foreach (Page<ConfigurationSetting> page in _options.PageableManager.GetPages(pageableSettings).ConfigureAwait(false))
                     {
-                        serverData[setting.Key] = setting;
+                        foreach (ConfigurationSetting setting in page.Values)
+                        {
+                            existingData[setting.Key] = setting;
+                        }
+
+                        matchConditions.Add(new MatchConditions { IfNoneMatch = page.GetRawResponse().Headers.ETag });
                     }
                 }).ConfigureAwait(false);
+
+                watchedCollections[new KeyValueIdentifier(KeyFilter.Any, LabelFilter.Null)] = matchConditions;
             }
 
-            foreach (KeyValueSelector loadOption in _options.KeyValueSelectors)
+            foreach (KeyValueSelector loadOption in selectors)
             {
-                IAsyncEnumerable<ConfigurationSetting> settingsEnumerable;
-
                 if (string.IsNullOrEmpty(loadOption.SnapshotName))
                 {
-                    settingsEnumerable = client.GetConfigurationSettingsAsync(
-                        new SettingSelector
+                    var selector = new SettingSelector()
+                    {
+                        KeyFilter = loadOption.KeyFilter,
+                        LabelFilter = loadOption.LabelFilter ?? LabelFilter.Null
+                    };
+
+                    List<MatchConditions> matchConditions = new List<MatchConditions>();
+
+                    await CallWithRequestTracing(async () =>
+                    {
+                        AsyncPageable<ConfigurationSetting> pageableSettings = client.GetConfigurationSettingsAsync(selector, cancellationToken);
+
+                        await foreach (Page<ConfigurationSetting> page in _options.PageableManager.GetPages(pageableSettings).ConfigureAwait(false))
                         {
-                            KeyFilter = loadOption.KeyFilter,
-                            LabelFilter = loadOption.LabelFilter
-                        },
-                        cancellationToken);
+                            foreach (ConfigurationSetting setting in page.Values)
```

Comment:
Can `page.Values` be null?

Reply:
I guess it could be? I don't see anything that explicitly says it won't be null, I just see that it returns an empty list normally when the query doesn't return any values

Reply:
I can add a null check

Reply:
What is the `page.Values` for 304 response?

Reply:
Empty list (IReadOnlyList)